pkg/jobrunaggregator/jobrunaggregatoranalyzer/pass_fail_test.go

```diff
@@ -0,0 +1,146 @@
+package jobrunaggregatoranalyzer
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+
+	"github.com/openshift/ci-tools/pkg/jobrunaggregator/jobrunaggregatorlib"
+)
+
+func createJobRunIDToAvailabilityResultForBackend(disruptions []int) map[string]jobrunaggregatorlib.AvailabilityResult {
+	jobRunIDToAvailabilityResultForBackend := make(map[string]jobrunaggregatorlib.AvailabilityResult)
+	for i, disruption := range disruptions {
+		jobRunIDToAvailabilityResultForBackend[fmt.Sprintf("run_%d", i)] = jobrunaggregatorlib.AvailabilityResult{SecondsUnavailable: disruption}
+	}
+
+	return jobRunIDToAvailabilityResultForBackend
+}
+
+func TestCheckPercentileDisruption(t *testing.T) {
+	weeklyAverageFromTenDays := weeklyAverageFromTenDays{}
+
+	tests := []struct {
+		name                 string
+		disruptions          []int
+		thresholdPercentile  int
+		historicalDisruption float64
+		status               testCaseStatus
+		failedCount          int
+		successCount         int
+	}{
+		{
+			// Required Passes for 95th percentile is 6
+			// no disruption so it passes
+			name:                 "Test 95th Percentile 0's Pass",
+			disruptions:          []int{0, 0, 0, 0, 0, 0, 0, 0, 0, 0},
+			thresholdPercentile:  95,
+			historicalDisruption: 7,
+			status:               testCasePassed,
+			failedCount:          0,
+			successCount:         10,
+		},
+		{
+			// Required Passes for 95th percentile is 6
+			// 5 Natural Passes
+			// Multiplier (6 - 5) == 1
+			// Fuzz Threshold ((historicalDisruption / 5) + 1) = 2
+			// The Disruption value that == 7 gets flipped to pass
```

Comment:
These tests are excellent, and especially the comments, could you update the logging to include these details as much as possible so we can see exactly what happened when we examine the test output. It looks like we'll log the threshold but the fuzziness details may help in debugging.

Reply:
How would the following be?  If there is not fuzziness applied it won't be included as in `run_1=0s` but for those cases, pass or fail, where fuzziness was evaluated the values used will be included in the test output `un_6=2s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 0.70) `

`Passed: Passed 5 times, failed 4 times.  (P85=1.30s requiredPasses=4 successes=[run_1=0s run_6=2s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 0.70) run_0=1s run_2=0s run_5=2s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 0.70)] failures=[run_3=5s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 3.70) run_4=5s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 3.70) run_7=3s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 1.70) run_8=4s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 2.70)])`

Reply:
Could we log anything constant to all earlier on (looks like multiplier and fuzzThreshold are constant), then trim down to just the fuzzValue in brackets with each failure.

And am I reading it right that for a failure, run_4=5s means my actual disruption was 5s, and the fuzzValue=3.70 is our forgiving value for it, which was still higher than 1.3s, so we fail.

Reply:
If we wanted the constants logged higher up only once AND only when the fuzziness logic was actually used I would have to return an additional flag, but that seems doable.

Almost, the 3.70 value is compared against the fuzzThreshold which is 1, not the historicalDisruptionValue which is the 1.30.

run_4=5s (Fuzzy Match: multiplier: 1.00, fuzzThreshold: 1.00, fuzzValue: 3.70)
Multiplier == 1,  We needed 4 passes but originally had only 3 so 4-3 = 1 == multiplier
fuzzThreshold = 1.00,  P85== 1.30, ((1.30 / 5) + 1) = 1 == fuzzThreshold) 
fuzzValue == 3.70, DisruptionValue - HistoricalValue * multiplier (( 5 - 1.30) * 1) == 3.70

3.70 > 1 so we still fail for this run.

Reply:
So yeah that is definitely cleaner :-)  let me know if you see a better way to format it.

`Passed: Passed 5 times, failed 4 times, with fuzzy matching values: multiplier 1.00, fuzzThreshold 1.00.  (P85=1.30s requiredPasses=4 successes=[run_0=1s run_1=0s run_2=0s run_5=2s (fuzzValue=0.70) run_6=2s (fuzzValue=0.70)] failures=[run_3=5s (fuzzValue=3.70) run_4=5s (fuzzValue=3.70) run_7=3s (fuzzValue=1.70) run_8=4s (fuzzValue=2.70)])`